agreement_dynamic/models/agreement_section.py

```diff
@@ -44,16 +51,29 @@ def onchange_copyvalue(self):
             self.copyvalue = "${{object.{}.{} or {}}}".format(
                 self.field_id.name,
                 self.sub_model_object_field_id.name,
-                self.default_value or "''",
+                self._get_proper_default_value(),
             )
 
     # compute the dynamic content for jinja expression
     def _compute_dynamic_content(self):
         MailTemplates = self.env["mail.template"]
         for this in self:
             content = MailTemplates._render_template(
-                this.content, "agreement.section", [this.id]
-            )[this.id]
-            this.dynamic_content = content
+                this.content, this.resource_ref_model_id.model, this.res_id,
+            )
+            this.dynamic_content = this._edit_expression(content)
 
+    def _get_proper_default_value(self):
+        self.ensure_one()
+        if not self.default_value:
+            return "''"
+        return "'{}'".format(self.default_value)
 
+    def _edit_expression(self, content):
+        """ Edit html with custom expressions"""
+        self.ensure_one()
+        for expression, replacement in POST_RENDERING_EXPRESSIONS.items():
+            if not content.find(expression):
+                continue
+            content = content.replace(expression, replacement)
+        return content
```

Comment:
What I had been hoping for is that we would be able to use the Mako rendering engine and just pass everything we need into the [variables](https://github.com/OCA/OCB/blob/13.0/addons/mail/models/mail_template.py#L292) before rendering, so that you could do

`${our_custom_placeholder_variable}`

It seems that there is no good way to inject new variables into the mail template rendering except by using the context, but then you'd always have to do `$(ctx['our_custom_placeholder_variable'])` which is really not very elegant.

Maybe you could directly call the Mako template rendering engine instead and then we can just inject all the variables you need there, and then we don't have to have two ways of injecting placeholders.

Reply:
That being said - if we are to do this we might just as well choose our own rendering engine, one that **doesnt** have the issue that the mail templates have where if you want to inject an "if" into the code all they layout gets shitfaced.

Reply:
TBH I don't know if such a templating language really exists. Maybe qweb itself, which is built on the idea of XML blocks, but I dont know if that will be any easier to edit in Summernote (the HTML editor) - i don't think so.

We're discouraging the use of IF anyway because we are building it in using the 'conditional blocks' idea - if they need more if's, they can just define more blocks.

So I would be fine if you go with the Mako/Jinja2 combo which Odoo uses for the mail templates. Should they have any trouble with the web editor we also have made a solution for that for another client which makes the 'underwater' HTML code view available to all users, and we could easily make that into an OCA module for them too.

Regarding mako/jinja2,  maybe use a SandboxedEnvironment - not sure what it is, but it sounds safe and good.

```
./mail/models/mail_template.py:34:    # We use a jinja2 sandboxed environment to render mako templates.
./mail/models/mail_template.py:41:    from jinja2.sandbox import SandboxedEnvironment
```

Reply:
OK! I didn't see an easy way to extend `mail.template._render_template()` to modify the `variables` it passes to Mako/Jinja.... and to fully override a function is bad practice, so that is why I suggested to copy relevant parts of that function into our own rendering function that calls the template engine directly.